src/Parallel/GlobalCache.hpp

```diff
@@ -193,9 +344,18 @@ class GlobalCache : public CBase_GlobalCache<Metavariables> {
               ParallelComponentTag>>&;  // NOLINT
 
   tuples::tagged_tuple_from_typelist<get_const_global_cache_tags<Metavariables>>
-      global_cache_{};
+      const_global_cache_{};
   tuples::tagged_tuple_from_typelist<parallel_component_tag_list>
       parallel_components_{};
+  // We store both a pointer and a proxy to the MutableGlobalCache.
+  // If the charm-aware constructor is used, then the pointer is set
+  // to nullptr and the proxy is set.
+  // If the non-charm-aware constructor is used, the the pointer is set
+  // and the proxy is ignored.
+  // The member functions that need the MutableGlobalCache should
+  // use the pointer if it is not nullptr, otherwise use the proxy.
+  MutableGlobalCache<Metavariables>* mutable_global_cache_{nullptr};
+  CProxy_MutableGlobalCache<Metavariables> mutable_global_cache_proxy_{};
```

Comment:
Yes this would move more checks to compile-time.  But I don't really like it because there would be `IsCharmAware` flying around throughout production code everywhere.  That problem would be solved by instead putting `IsGlobalCacheCharmAware` in the Metavariables (which GlobalCache already knows about).  Not sure if it's worth it.  Again, I'd welcome others' thoughts too.

Reply:
Yeah, I agree with these concerns. Not a change you should make unless there was a solid consensus :)